File: runtime/html-rewriter.mjs
// Cloudflare-Workers-shape `HTMLRewriter` global for Node.js, backed by a WASM
// build of lol-html (Cloudflare's streaming HTML rewriter). Code written against
// Cloudflare Workers' (or Bun's) HTMLRewriter ports here unchanged:
//
//   new HTMLRewriter()
//     .on("a[href]", { element(el) { el.setAttribute("rel", "noopener"); } })
//     .transform(response)
//
// ENGINE: the vendored html-rewriter-wasm build (runtime/html-rewriter-engine/),
// lol-html compiled to WebAssembly with Binaryen's Asyncify pass. One portable
// .wasm for every platform — no per-platform native prebuilds. Asyncify is what
// gives FULL ASYNC-HANDLER support: when a handler returns a Promise, the engine
// unwinds the WASM stack, awaits the Promise, then rewinds and continues the
// transform — so `engine.write()`/`engine.end()` are async and a handler may be
// `async`/return a Promise (Cloudflare/Bun parity). The engine owns parsing + the
// rewritable-unit methods; this file is the thin fluent wrapper + the WHATWG
// Response streaming bridge.

// node: builtins via process.getBuiltinModule (NOT static import) — the same loader-
// chain-leak avoidance the Worker polyfill documents: a static `import` would route
// the builtin through the user's --loader/register() hooks. On the floor where
// getBuiltinModule is absent, createRequire is threaded in via the setter below.
let _bootstrapCreateRequire = null;
export function setBootstrapCreateRequire(fn) {
  _bootstrapCreateRequire = fn;
}
function __getBuiltin(id) {
  if (typeof process.getBuiltinModule === "function") return process.getBuiltinModule(id);
  return _bootstrapCreateRequire(import.meta.url)(id);
}

// Resolve the WASM engine constructor lazily + memoized. The engine is a CommonJS
// module in nub's distribution (runtime/html-rewriter-engine/html_rewriter.js); it
// instantiates the .wasm synchronously off its own __dirname, so it never touches
// the ESM loader chain. Loaded only on the first transform — non-HTMLRewriter runs
// never pay the ~900KB .wasm instantiation.
let _engineCtor;
function getEngineCtor() {
  if (_engineCtor !== undefined) return _engineCtor;
  const { createRequire } = __getBuiltin("node:module");
  const { fileURLToPath } = __getBuiltin("node:url");
  const require = createRequire(import.meta.url);
  _engineCtor = null;
  for (const rel of [
    "./html-rewriter-engine/html_rewriter.js",
    "../runtime/html-rewriter-engine/html_rewriter.js",
  ]) {
    try {
      const mod = require(fileURLToPath(new URL(rel, import.meta.url)));
      if (mod && mod.HTMLRewriter) {
        _engineCtor = mod.HTMLRewriter;
        break;
      }
    } catch {
      // try the next candidate path
    }
  }
  return _engineCtor;
}

// Output sink for the throwaway engine used only to validate selectors at .on().
const NOOP_SINK = () => {};

function requireEngine() {
  const Engine = getEngineCtor();
  if (!Engine) {
    throw new Error(
      "HTMLRewriter: the WASM engine is unavailable (html-rewriter-engine not found).",
    );
  }
  return Engine;
}

function assertHandlers(handlers) {
  if (handlers == null || typeof handlers !== "object") {
    throw new TypeError("HTMLRewriter: handlers must be an object");
  }
  return handlers;
}

class HTMLRewriter {
  // Registrations are buffered until transform(): a fresh engine is built per
  // transform (the WASM engine is single-use — one end() per instance), so one
  // HTMLRewriter can transform multiple inputs (Cloudflare parity).
  #elementHandlers = [];
  #documentHandlers = [];

  on(selector, handlers) {
    if (typeof selector !== "string") {
      throw new TypeError("HTMLRewriter.on: selector must be a string");
    }
    assertHandlers(handlers);
    // Validate the selector eagerly so an invalid selector throws HERE, matching
    // Cloudflare's "throws at .on() registration" contract. The real engine is
    // built per-transform, but a throwaway engine parses the selector immediately
    // and surfaces the error now. free() it so the WASM instance isn't leaked.
    const Engine = getEngineCtor();
    if (Engine) {
      const probe = new Engine(NOOP_SINK);
      try {
        probe.on(selector, {});
      } finally {
        probe.free();
      }
    }
    this.#elementHandlers.push([selector, handlers]);
    return this;
  }

  onDocument(handlers) {
    assertHandlers(handlers);
    this.#documentHandlers.push(handlers);
    return this;
  }

  #buildEngine(sink) {
    const Engine = requireEngine();
    const engine = new Engine(sink);
    for (const [selector, h] of this.#elementHandlers) engine.on(selector, h);
    for (const h of this.#documentHandlers) engine.onDocument(h);
    return engine;
  }

  // Cloudflare-exact: transform a Response, returning a new streaming Response.
  transform(input) {
    if (!(input instanceof Response)) {
      throw new TypeError("HTMLRewriter.transform: input must be a Response");
    }

    const sourceBody = input.body;
    if (sourceBody == null) {
      // No body to rewrite — return an equivalent empty-body Response.
      return new Response(null, input);
    }

    const self = this;
    let engine = null;
    let reader = null;
    // Set when the consumer cancels the output stream. The source-read loop bails
    // and skips end()/close() so it can't write/enqueue into a torn-down stream.
    let cancelled = false;

    // Free the WASM engine, tolerating a held wasm-bindgen borrow. If the consumer
    // cancels WHILE an async handler is suspended (Asyncify), the suspended Rust
    // write frame still holds its RefCell borrow, so free() throws "recursive use
    // of an object". Swallowing that here is safe: the suspended write resumes when
    // its handler promise settles, its frame returns and drops the borrow, and
    // start()'s `finally{cleanup()}` then frees the engine on that resume. Without
    // the guard the throw escapes reader.cancel() (rejecting an ordinary abort) AND
    // leaks the engine.
    const safeFree = () => {
      if (!engine) return;
      try {
        engine.free();
        engine = null;
      } catch (_) {
        // Borrow still held (cancel-mid-suspend); freed on resume — see above.
      }
    };

    // Release the source reader + free the engine. Idempotent.
    const cleanup = () => {
      if (reader) {
        try {
          reader.releaseLock();
        } catch (_) {
          // already released (e.g. by reader.cancel in the cancel path)
        }
        reader = null;
      }
      safeFree();
    };

    const stream = new ReadableStream({
      async start(controller) {
        engine = self.#buildEngine((chunk) => {
          // The engine hands back a Uint8Array view into WASM memory; copy it,
          // since the buffer is reused across chunks. Skip once cancelled — the
          // controller is torn down.
          if (!cancelled && chunk && chunk.length) controller.enqueue(new Uint8Array(chunk));
        });
        reader = sourceBody.getReader();
        try {
          for (;;) {
            if (cancelled) break;
            const { done, value } = await reader.read();
            if (done || cancelled) break;
            // Async: the engine awaits any Promise a handler returns (Asyncify).
            await engine.write(value);
          }
          if (!cancelled) {
            await engine.end();
            controller.close();
          }
        } catch (err) {
          if (!cancelled) controller.error(err);
        } finally {
          cleanup();
        }
      },
      // Consumer aborted the output stream: stop reading the source body and let
      // start()'s `finally{cleanup()}` own ALL freeing. cancel() must NOT call
      // free() itself: if an async handler is in flight the write is suspended with
      // the Rust borrow held, so free() throws — but wasm-bindgen's
      // __destroy_into_raw() has ALREADY zeroed the wrapper's ptr before the
      // throwing wasm free, orphaning a still-alive Rust object (ptr=0, object
      // live). cleanup() then can only retry on a null pointer → the engine leaks
      // permanently and the shared WASM memory eventually OOBs. Deferring all
      // freeing to cleanup() — which runs only AFTER the suspended write resumes
      // and the borrow releases — frees exactly once, cleanly.
      cancel(reason) {
        cancelled = true;
        const r = reader;
        // r.cancel(reason) releases the reader lock; capture its promise first.
        const c = r ? r.cancel(reason) : undefined;
        reader = null;
        return c;
      },
    });

    // Rewriting changes byte length, so content-length must not carry over.
    const headers = new Headers(input.headers);
    headers.delete("content-length");
    return new Response(stream, {
      status: input.status,
      statusText: input.statusText,
      headers,
    });
  }
}

export function installHTMLRewriter() {
  // Forward-compat / additive: if a real HTMLRewriter is already present (a future
  // Node, or the user's own), do nothing.
  if (typeof globalThis.HTMLRewriter !== "undefined") return;
  // NON-ENUMERABLE: invisible to Object.keys(globalThis)/for-in is the additive
  // contract — vanilla-Node code enumerating the global object must not observe
  // nub's injected global. writable+configurable so user code can override it.
  Object.defineProperty(globalThis, "HTMLRewriter", {
    value: HTMLRewriter,
    enumerable: false,
    writable: true,
    configurable: true,
  });
}

// Fast tier (getBuiltinModule present): install eagerly at module eval, preserving
// the side-effect-on-require contract the lazy preload getter relies on. The engine
// (and its .wasm) is still resolved lazily on the first transform, so this costs
// ~nothing for the common "never touch HTMLRewriter" run. On the floor the compat
// preload calls setBootstrapCreateRequire(...) + installHTMLRewriter() explicitly.
if (typeof process.getBuiltinModule === "function") installHTMLRewriter();

File: crates/nub-cli/tests/html_rewriter.rs
//! HTMLRewriter integration tests: run fixture scripts through the built `nub`
//! binary and assert the rewrite output. Exercises the Cloudflare-Workers-shape
//! global backed by the vendored WASM lol-html engine
//! (`runtime/html-rewriter-engine/`, lol-html compiled to WebAssembly with
//! Asyncify), which ships in nub's distribution — no native addon needed for
//! HTMLRewriter. Covers the API surface, async handlers (awaited mid-transform),
//! async-handler rejection, consumer cancel (incl. mid-suspend), the engine
//! leak-loop, and compat-mode absence.

use std::path::{Path, PathBuf};
use std::process::Command;

fn nub_binary() -> PathBuf {
    let mut path = std::env::current_exe().unwrap();
    path.pop(); // deps/
    path.pop(); // debug/
    path.push(format!("nub{}", std::env::consts::EXE_SUFFIX));
    path
}

fn fixture(file: &str) -> PathBuf {
    let manifest = std::env::var("CARGO_MANIFEST_DIR").unwrap();
    Path::new(&manifest)
        .join("../../tests/fixtures/html-rewriter")
        .join(file)
}

fn run(file: &str, extra_args: &[&str], env: &[(&str, &str)]) -> (String, String, i32) {
    let mut cmd = Command::new(nub_binary());
    for a in extra_args {
        cmd.arg(a);
    }
    cmd.arg(fixture(file));
    for &(k, v) in env {
        cmd.env(k, v);
    }
    let out = cmd.output().expect("failed to spawn nub");
    (
        String::from_utf8_lossy(&out.stdout).to_string(),
        String::from_utf8_lossy(&out.stderr).to_string(),
        out.status.code().unwrap_or(-1),
    )
}

/// The full surface: the global is present + non-enumerable, attribute/content
/// mutation, escaped-vs-raw insertion, element removal, document-end append,
/// doctype read, text rewriting, async handlers awaited mid-transform, streaming
/// over a Response, the non-Response TypeError, and the synchronous selector-error.
#[test]
fn rewrites_html_across_the_workers_api_surface() {
    let (stdout, stderr, code) = run("rewrite.mjs", &[], &[]);
    assert_eq!(code, 0, "fixture must exit 0\nstderr: {stderr}");

    // Attribute injection + setInnerContent.
    assert!(
        stdout.contains(r#"ATTR: <h1>Hi</h1><a href="/" rel="noopener">link</a>"#),
        "attribute/content rewrite wrong:\n{stdout}"
    );
    // Raw HTML inserted verbatim; text-mode insertion HTML-escaped.
    assert!(
        stdout.contains("CONTENT: <p>x<b>raw</b>&lt;i&gt;esc&lt;/i&gt;</p>"),
        "escaped-vs-raw insertion wrong:\n{stdout}"
    );
    // Doctype name is readable.
    assert!(
        stdout.contains("DOCTYPE: html"),
        "doctype name wrong:\n{stdout}"
    );
    // Element removed; document-end content appended.
    assert!(
        stdout.contains("REMOVE: <!DOCTYPE html><div>keep</div><!--end-->"),
        "remove + document-end append wrong:\n{stdout}"
    );
    // Text rewriting.
    assert!(
        stdout.contains("TEXT: <span>HELLO</span>"),
        "text rewrite wrong:\n{stdout}"
    );
    // Async (Promise-returning) handlers are awaited mid-transform (Asyncify):
    // the element handler awaits before setAttribute, the document-end handler
    // awaits before append, and both land in the output.
    assert!(
        stdout.contains(r#"ASYNC: <a href="/" data-async="1">x</a><!--async-end-->"#),
        "async handlers must be awaited mid-transform:\n{stdout}"
    );
    // Streaming over a Response yields the transformed body.
    assert!(
        stdout.contains("STREAM: <title>Streamed</title>"),
        "Response streaming wrong:\n{stdout}"
    );
    // Cloudflare-exact: a non-Response input throws a TypeError.
    assert!(
        stdout.contains("BADINPUT: true"),
        "non-Response transform input must throw:\n{stdout}"
    );
    // Invalid selector throws synchronously at .on().
    assert!(
        stdout.contains("BADSEL: true"),
        "invalid selector must throw:\n{stdout}"
    );

    assert!(
        stdout.contains("DONE"),
        "fixture did not run to completion:\n{stdout}"
    );
}

/// A rejecting async handler must propagate the ORIGINAL error (not the secondary
/// borrow/aliasing error the Asyncify rewind would surface), and a subsequent
/// transform on a fresh instance must still work. Guards the asyncify.js
/// rewind-on-reject patch.
#[test]
fn async_handler_rejection_propagates_original_error() {
    let (stdout, stderr, code) = run("async-reject.mjs", &[], &[]);
    assert_eq!(code, 0, "fixture must exit 0\nstderr: {stderr}");
    assert!(
        stdout.contains("REJECT_ORIGINAL: true"),
        "rejecting async handler must surface the ORIGINAL error:\n{stdout}"
    );
    assert!(
        stdout.contains("REJECT_RECOVERS: true"),
        "a fresh transform after a rejection must still work:\n{stdout}"
    );
    assert!(
        stdout.contains("DONE"),
        "fixture did not complete:\n{stdout}"
    );
}

/// Cancelling the transformed output stream must resolve cleanly — including when
/// an async handler is suspended mid-transform (the held WASM borrow is tolerated
/// and the engine is freed on resume). Guards the safeFree + cancelled-flag fix.
#[test]
fn consumer_cancel_resolves_even_mid_suspend() {
    let (stdout, stderr, code) = run("cancel.mjs", &[], &[]);
    assert_eq!(code, 0, "fixture must exit 0\nstderr: {stderr}");
    assert!(
        stdout.contains("CANCEL_PLAIN_RESOLVED: true"),
        "plain consumer cancel must resolve:\n{stdout}"
    );
    assert!(
        stdout.contains("CANCEL_MIDSUSPEND_RESOLVED: true"),
        "cancel during an in-flight async handler must resolve (not reject):\n{stdout}"
    );
    assert!(
        stdout.contains("DONE"),
        "fixture did not complete:\n{stdout}"
    );
}

/// The cancel-mid-suspend path must RECLAIM the engine: a pure cancel-mid-suspend
/// loop must neither crash (the pre-fix path OOB-crashed by ~1000 cycles) nor grow
/// RSS across a post-warmup measure window (a per-engine WASM leak would). This is
/// the residual-leak guard — cancel() must defer all freeing to cleanup() so it
/// frees exactly once, after the suspended write resumes and the borrow releases.
#[test]
fn cancel_mid_suspend_reclaims_the_engine() {
    // --expose-gc gives a stable RSS reading; nub forwards V8 flags to Node.
    let (stdout, stderr, code) = run("leak-loop.mjs", &["--expose-gc"], &[]);
    // exit 0 + reaching the markers means no "memory access out of bounds" crash.
    assert_eq!(
        code, 0,
        "fixture must exit 0 (no OOB crash)\nstderr: {stderr}"
    );
    assert!(
        stdout.contains("CANCEL_MIDSUSPEND_NO_CRASH: true"),
        "cancel-mid-suspend loop must not crash:\n{stdout}"
    );
    assert!(
        stdout.contains("CANCEL_MIDSUSPEND_FLAT: true"),
        "cancel-mid-suspend must reclaim the engine (RSS flat across the window):\n{stdout}"
    );
    assert!(
        stdout.contains("DONE"),
        "fixture did not complete:\n{stdout}"
    );
}

/// Under `--node` (zero augmentation) the global is absent.
#[test]
fn absent_under_node_flag() {
    let (stdout, stderr, code) = run("compat-absent.mjs", &["--node"], &[]);
    assert_eq!(code, 0, "stderr: {stderr}");
    assert!(
        stdout.contains("HTMLREWRITER: undefined"),
        "HTMLRewriter must be absent under --node:\n{stdout}"
    );
}

/// A truthy `NODE_COMPAT` is the persistent tree-wide augmentation opt-out — same
/// effect as `--node`.
#[test]
fn absent_under_node_compat_env() {
    let (stdout, stderr, code) = run("compat-absent.mjs", &[], &[("NODE_COMPAT", "1")]);
    assert_eq!(code, 0, "stderr: {stderr}");
    assert!(
        stdout.contains("HTMLREWRITER: undefined"),
        "HTMLRewriter must be absent under NODE_COMPAT=1:\n{stdout}"
    );
}
